lisp/ess-inf.el

```diff
@@ -2811,9 +2812,20 @@ To be used in `ess-idle-timer-functions'."
   ;; default-directory and subprocess working directory are
   ;; synchronized automatically.
   (interactive)
-  (let ((dir (car (ess-get-words-from-vector "getwd()\n"))))
+  (let* ((dir (car (ess-get-words-from-vector "getwd()\n")))
+         (new-default-dir (ess-path-update-local-portion default-directory dir)))
     (message "(ESS / default) directory: %s" dir)
-    (setq default-directory (file-name-as-directory dir))))
+    (setq default-directory (file-name-as-directory new-default-dir))))
+
+(defun ess-path-update-local-portion (current-path new-local-path)
```

Comment:
This function should be private. I think this signature might be clearer:

```suggestion
(defun ess--derive-connection-path (old new)
```

The docstring could be shortened to simply mention that a new connection directory is derived from `OLD` (a path) and `NEW` (a path), in such a way that the host and connection information in `OLD` is retained in the `NEW` path. `NEW` must be an absolute local path, it can't contain connection information.

Reply:
I've updated the function signature and the docstring. Your comments made me think a little bit more about the function accepting only a local path for the input to `new`, and so I've updated the function so that this can also be a remote path.